### PR TITLE
Adding layerwise folding as tutorial to mitiq.

### PR DESCRIPTION
**Pull request by vprusso**
- Tutorial provided to showcase functionality for layerwise folding and error mitigation.

### CODE REVIEW

**Comment by vprusso** [comment]
Not sure if I'm missing something, but all of the tests within the `build/docs` directory appear to be passing (link), but somehow the whole docs pipeline fails. Is there an obvious reason as to why this might be the case?

File: docs/source/examples/layerwise-folding.md [hunk 1]
```diff
@@ -0,0 +1,305 @@
...
+    format_name: myst
+    format_version: 0.13
+    jupytext_version: 1.11.1
+kernelspec:
+  display_name: Python 3
+  language: python
+  name: python3
+---
+
+# Layerwise folding in Mitiq
```
Comment (nathanshammah) [inline docs/source/examples/layerwise-folding.md:14]:
```suggestion
# ZNE with Qiskit: Layerwise folding
```

**Comment by nathanshammah** [comment]
Thanks @vprusso! This is a great addition to the documentation. 
Regarding the failing tests, it looks like CI doesn't like [#step:6:349](https://github.com/unitaryfund/mitiq/actions/runs/5429536082/jobs/9875012613?pr=1894#step:6:349), 
```
/home/runner/work/mitiq/mitiq/docs/source/examples/layerwise-folding.md:25: WARNING: 'myst' reference target not found: total-variational-distance-metric

/home/runner/work/mitiq/mitiq/docs/source/examples/layerwise-folding.md:26: WARNING: 'myst' reference target not found: impact-of-single-vs.-double-folding
/home/runner/work/mitiq/mitiq/docs/source/examples/layerwise-folding.md:28: WARNING: 'myst' reference target not found: #experiment:-global-folding-with-linear-extrapolation
/home/runner/work/mitiq/mitiq/docs/source/examples/layerwise-folding.md:29: WARNING: 'myst' reference target not found: #experiment:-layerwise-folding-with-linear-extrapolation
```
Can you please add the file to the index file, at [examples.md](https://github.com/unitaryfund/mitiq/blob/master/docs/source/examples/examples.md)? It would be great to include a thumbnail too, including a path for it in the conf.py and adding it to the `_thumbnails` folder. You can check the correct build of the docs here in the CI test list, looking at the [RTD build](https://mitiq--1894.org.readthedocs.build/en/1894/).

**Comment by vprusso** [comment]
Awesome. Thanks, @nathanshammah. We should be all ✅ now!

File: docs/source/examples/layerwise-folding.md [hunk 2]
```diff
@@ -0,0 +1,305 @@
...
+    noise_model = None
+else:
+    # Default to a simulator.
+    backend = qiskit.Aer.get_backend("qasm_simulator")
+    noise_model = initialized_depolarizing_noise(noise_level=0.02)
+
+shots = 10_000
+```
+
+**Note:** When `USE_REAL_HARDWARE` is set to `False`, a classically simulated noisy backend is used instead of a real quantum computer.
```
Comment (andreamari) [inline docs/source/examples/layerwise-folding.md:70]: Not sure if we need to support the `USE_REAL_HARDWARE` option in this example. We can remove this part and just keep:
```
backend = qiskit.Aer.get_backend("qasm_simulator")
noise_model = initialized_depolarizing_noise(noise_level=0.02)
```

File: docs/source/examples/layerwise-folding.md [hunk 3]
```diff
@@ -0,0 +1,305 @@
...
+that corresponds to measuring bit string $k$. 
+
+The *impact* of applying an inversion is given by
+
+$$
+d \left[p(\cdot|C), p(\cdot|C^{(i)})\right]
+$$
+
+where $d$ is some distance measure. In
+[arXiv:2204.06056](https://arxiv.org/abs/2204.06056), the authors used the
```
Comment (andreamari) [inline docs/source/examples/layerwise-folding.md:161]: I would cite the bib item of the same paper. We can still keep also the explicit arxiv link, if you think it is useful.

**Review by andreamari — CHANGES_REQUESTED** [review]
Thanks Vincent, very nice example!
Minor suggestions are given below.

File: docs/source/examples/layerwise-folding.md [hunk 4]
```diff
@@ -0,0 +1,305 @@
...
+  language: python
+  name: python3
+---
+
+# ZNE with Qiskit: Layerwise folding
+
+
+This tutorial shows an example of how to mitigate noise on IBMQ backends with
+the Cirq frontend using layerwise folding in contrast with global folding.
+
```
[the remark this answers is not in the document]
Comment (vprusso) [reply docs/source/examples/layerwise-folding.md:19]: Good call! Done.

File: docs/source/examples/layerwise-folding.md (hunk 2 above)
Comment (vprusso) [reply docs/source/examples/layerwise-folding.md:70]: Yeah, having that extra option is nice, but it kind of obscures the main theme of the intent of this tutorial. Removed!

File: docs/source/examples/layerwise-folding.md (hunk 3 above)
Comment (vprusso) [reply docs/source/examples/layerwise-folding.md:161]: Cited here as cited elsewhere in the ZNE section. Thanks!

File: docs/source/examples/layerwise-folding.md [hunk 5]
```diff
@@ -0,0 +1,305 @@
...
+```{code-cell} ipython3
+# Define a basic circuit for
+q0, q1 = cirq.LineQubit.range(2)
+circuit = cirq.Circuit(
+    [cirq.ops.H(q0)],
+    [cirq.ops.CNOT(q0, q1)],
+    [cirq.measure(cirq.LineQubit(0))],
+)
+draw(circuit)
```
[the remark this answers is not in the document]
Comment (vprusso) [reply docs/source/examples/layerwise-folding.md:103]: That's odd. Removed the `draw` method and replaced all calls with `print` instead. Thank you!

File: docs/source/examples/layerwise-folding.md [hunk 6]
```diff
@@ -0,0 +1,305 @@
...
+
+Define the quantity:
+
+$$
+p(k|C) = \langle \langle k | C | \rho_0 \rangle \rangle
+$$
+
+as the probability distribution oveer measurement outcomes at the output of a
+circuit $C$ where $k \in B^n$ with $B^n$ being the set of all $n$-length bit
+strings where $\langle \langle k |$ is the vectorized (non-ideal) POVM element
```
[the remark this answers is not in the document]
Comment (vprusso) [reply docs/source/examples/layerwise-folding.md:151]: I took this language from the cited paper: https://arxiv.org/pdf/2204.06056.pdf (Appendix-A). Even with the surrounding context, I'm actually still not sure what "non-ideal" here means. I'll go ahead and take it out for now, but if it makes sense to put it back in with more context, I can do that as well. Thanks!

**Comment by vprusso** [comment]
Thanks again for the comments, @andreamari ! 

These should all (hopefully) be addressed now. Thank you!

**Review by andreamari — APPROVED** [review]

**Comment by vprusso** [comment]
Thanks for the approval, @andreamari. Any qualms @natestemen , or would you say this is good to go?

File: docs/source/examples/layerwise-folding.md [hunk 7]
```diff
@@ -0,0 +1,289 @@
+---
+jupytext:
+  text_representation:
+    extension: .myst
```
Comment (natestemen) [inline docs/source/examples/layerwise-folding.md:4]:
```suggestion
    extension: .md
```

File: docs/source/examples/layerwise-folding.md [hunk 8]
```diff
@@ -0,0 +1,289 @@
...
+fold_layer_func = zne.scaling.get_layer_folding(layer_to_fold)
+
+mitigated = zne.execute_with_zne(circuit, executor, scale_noise=fold_layer_func, factory=linear_factory)
+print(f"Mitigated (layerwise folding) result {mitigated:.3f}")
+print(f"Unmitigated result {unmitigated:.3f}")
+```
+
+**Note**: While doing layerwise folding on the noisiest layer will, on average,
+improve the mitigated value, it still will not eclipse the benefit of doing
+global folding.
```
Comment (natestemen) [inline docs/source/examples/layerwise-folding.md:280]: Might be better to use MyST's `note` directive:
````suggestion
```{note}
While doing layerwise folding on the noisiest layer will, on average,
improve the mitigated value, it still will not eclipse the benefit of doing
global folding.
```
````

File: docs/source/examples/layerwise-folding.md [hunk 9]
```diff
@@ -0,0 +1,289 @@
...
+
+So why consider this technique in this context? One reason is that applying
+global folding will increase the length of the entire circuit while layerwise
+folding on a subset of only the noisiest layers will increase the circuit by a
+smaller factor. 
+
+If running a circuit on hardware is bottle-necked by the cost of running a long
+circuit, this technique could potentially be used to arrive at a better result
+(although not as good as global folding) but with less monetary cost.
```
Comment (natestemen) [inline docs/source/examples/layerwise-folding.md:289]: I wonder if this motivational material should be moved to the top?

**Review by natestemen — COMMENTED** [review]
Nice work Vincent, thanks for putting this together.

Biggest thing is just a minor comment about moving some of the motivational context to the beginning of the tutorial.

File: docs/source/examples/layerwise-folding.md (hunk 9 above)
Comment (vprusso) [reply docs/source/examples/layerwise-folding.md:300]: Good call. Done!

**Comment by vprusso** [comment]
Thanks for the comments, @natestemen ! 

I believe they have been addressed. Let me know if there's anything else, but if not, I think it's good to go!

File: docs/source/examples/layerwise-folding.md (hunk 9 above)
Comment (vprusso) [reply docs/source/examples/layerwise-folding.md]: Nope, I'm just a fool who left it in there twice. Removed, and thank you for the catch!

**Review by natestemen — APPROVED** [review]
:shipit:

**Comment by vprusso** [comment]
Awesome, thanks for the approval, @natestemen. I believe things are good (although it seems like the docs are failing for some odd reason). I know that they sometimes do this, so I'm not sure if this is just an artifact of something unrelated. 

I attempted to rerun the docs build, but it still seems to fail. It mentions something about bsqkit (which I didn't touch in my MR):

```
/home/runner/work/mitiq/mitiq/docs/source/examples/bqskit.md: WARNING: Executing notebook failed: CellExecutionError [mystnb.exec]
/home/runner/work/mitiq/mitiq/docs/source/examples/bqskit.md: WARNING: Notebook exception traceback saved in: /home/runner/work/mitiq/mitiq/docs/build/html/reports/examples/bqskit.err.log [mystnb.exec]
```

Is it okay to merge or should I hold off?

**Comment by natestemen** [comment]
Yup that's a problem that we are seeing across all PRs are the moment, so no need to worry about that. Sorry for the confusion. I'll let you hit the merge button when you want since it's fun :)

**Comment by vprusso** [comment]
It *is* fun! :) Away it goes 🚀

**Pull request merged**